dev/ci/ci-common.sh

```diff
@@ -46,6 +46,40 @@ done
 
 mathcomp_CI_DIR="${CI_BUILD_DIR}/math-comp"
 
+# [git_download project dest] will download the given [ref] of [project]
+# and unpack it in [dest] if the folder does not exist already;
+# if it does, it will do nothing except print a warning
+# (this can be useful when testing make ci-* locally).
+# When the variable $CI is undefined, it uses git clone to perform the
+# download.
+git_download()
+{
+  local PROJECT=$1
+  local DEST=$2
+
+  if [ -d "$DEST" ]; then
+    echo "Warning: download and unpacking of $PROJECT skipped because $DEST already exists."
+  elif [ -n "$CI" ]; then # When on CI, we download tarballs
+    local ARCHIVEURL_VAR="${PROJECT}_CI_ARCHIVEURL"
+    local ARCHIVEURL="${!ARCHIVEURL_VAR}"
+    local REF_VAR="${PROJECT}_CI_REF"
+    local REF="${!REF_VAR}"
+    mkdir -p "$DEST"
+    cd "$DEST"
+    wget "$ARCHIVEURL/$REF.tar.gz"
+    tar xvfz "$REF.tar.gz" --strip-components=1
+    rm -f "$REF.tar.gz"
+  else # When on local machine, we clone git repo to help prepare overlays
+    local GITURL_VAR="${PROJECT}_CI_GITURL"
+    local GITURL="${!GITURL_VAR}"
+    local REF_VAR="${PROJECT}_CI_REF"
+    local REF="${!REF_VAR}"
+    git clone "$GITURL" "$DEST"
```

Comment:
Could you replicate the stuff in `git_checkout` where it still works if the directory already exists?
Or perhaps even call the `git_checkout` function here.

Reply:
I've considered this but my past experience with the way `git_checkout` behaves is far from satisfying so that's why I've preferred to aim for a simpler approach: if the directory already exists either delete it or manually checkout the right branch yourself.
It will also avoid any problem if the data was previously fetched as a tarball (if you had defined `$CI`).
Does this sound OK?

Reply:
> Does this sound OK?

No, it means I can't do `make ci-foo` after fixing some file in foo to see if it worked.

Reply:
> my past experience with the way git_checkout behaves is far from satisfying

Can you say more? A key goal I have is actually remove all custom build rules for ci projects as they move to Dune [then for example if mathcomp depends on coq, dune will use the local coq and track all the global deps]

So IMHO it is important to support @SkySkimmer workflow.

Reply:
OK I think I had misunderstood @SkySkimmer's initial comment so answering differently:
`git_download` will not fail if the directory already exists, it will just print a warning and skip directly to the build phase. So your workflow (a very important one indeed) is fully supported.